flow-report/src/common.tsx

```diff
@@ -7,25 +7,62 @@
 import {FunctionComponent} from 'preact';
 
 import {NavigationIcon, SnapshotIcon, TimespanIcon} from './icons';
+import {getScreenDimensions, getScreenshot} from './util';
 
 export const Separator: FunctionComponent = () => {
   return <div className="Separator" role="separator"></div>;
 };
 
+export const FlowStepIcon: FunctionComponent<{mode: LH.Result.GatherMode}> = ({mode}) => {
+  return <>
+    {
+      mode === 'navigation' && <NavigationIcon/>
+    }
+    {
+      mode === 'timespan' && <TimespanIcon/>
+    }
+    {
+      mode === 'snapshot' && <SnapshotIcon/>
+    }
+  </>;
+};
+
 export const FlowSegment: FunctionComponent<{mode?: LH.Result.GatherMode}> = ({mode}) => {
   return (
     <div className="FlowSegment">
       <div className="FlowSegment__top-line"/>
       {
-        mode === 'navigation' && <NavigationIcon/>
-      }
-      {
-        mode === 'timespan' && <TimespanIcon/>
-      }
-      {
-        mode === 'snapshot' && <SnapshotIcon/>
+        mode && <FlowStepIcon mode={mode}/>
       }
       <div className="FlowSegment__bottom-line"/>
     </div>
   );
 };
+
+export const FlowStepThumbnail: FunctionComponent<{
+  reportResult: LH.ReportResult,
+  width?: number,
+  height?: number,
+}> = ({reportResult, width, height}) => {
+  const screenshot = getScreenshot(reportResult);
+
+  // Crop the displayed image to the viewport dimensions.
+  const dimensions = getScreenDimensions(reportResult);
+  if (width && height === undefined) {
+    height = dimensions.height * width / dimensions.width;
+  } else if (height && width === undefined) {
+    width = dimensions.width * height / dimensions.height;
+  }
```

Comment:
wondering if the new css `aspect-ratio` would be a good use here?

Reply:
I have been avoiding it because I thought it was unsupported in Safari, but looks like support was recently added https://caniuse.com/?search=aspect-ratio. We would still be using JS to set style property though.